sycl/include/sycl/ext/oneapi/properties/property.hpp

```diff
@@ -195,8 +195,10 @@ enum PropKind : uint32_t {
   UsesValid = 29,
   UseRootSync = 30,
   RegisterAllocMode = 31,
+  GRFSize = 32,
+  GRFSizeAutomatic = 33,
```

Comment:
What are we currently doing to prevent the user having `grf_size` and `grf_size_automatic` in the same property list? One option would be to let them have the same PropKind which should cause it to complain about "duplicate" properties. However, that might also be a confusing error for this case, since they aren't the same per se.

Alternatively, we could add a new trait for mutually exclusive properties and have `properties` assert that there are no mutually exclusive properties. For performance it might be best to have something like:

```c++
// Specializations of the following trait should not consider itself a conflicting property.
template<typename PropKey, typename Properties> struct HasConflictingProp : std::false_type {};

template<typename Properties> struct HasConflictingProp<grf_size_key, Properties>
  : std::bool_constant<Properties::has_property<grf_size_automatic_key>()> {};

template<typename Properties> struct HasConflictingProp<grf_size_automatic_key, Properties>
  : std::bool_constant<Properties::has_property<grf_size_key>()> {};
```

Then you would make another trait for iterating the keys in the properties together with the property list. Note that we technically don't need both of the specializations above, but for completeness sake we might as well.

Reply:
Yeah this is a great point, in the spec we explicitly say doing this is undefined, but we should have an error because otherwise the user experience is really bad. I like your second idea, let me try to use that example and get it working.

Reply:
@steffenlarsen I attempted to add this in my latest commit, but I got really close to hitting my limits with metaprogramming, so I would appreciate your review. It's very possible I missed a case or did something in a roundabout way. Thanks.

Reply:
Looks great, thanks! And an extra 👍  for adding tests. 😄